handle.go

```diff
@@ -0,0 +1,184 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+//go:build (linux || darwin) && (amd64 || arm64)
+
+package waf
+
+import (
+	"errors"
+	"fmt"
+	emptyfree "github.com/DataDog/go-libddwaf/internal/empty-free"
+	"go.uber.org/atomic"
+	"reflect"
+	"sync"
+)
+
+// Handle represents an instance of the WAF for a given ruleset.
+type Handle struct {
+	// Instance of the WAF
+	cHandle wafHandle
+
+	// Lock-less reference counter avoiding blocking calls to the Close() method
+	// while WAF contexts are still using the WAF handle. Instead, we let the
+	// release actually happen only when the reference counter reaches 0.
+	// This can happen either from a request handler calling its WAF context's
+	// Close() method, or either from the appsec instance calling the WAF
+	// handle's Close() method when creating a new WAF handle with new rules.
+	// Note that this means several instances of the WAF can exist at the same
+	// time with their own set of rules. This choice was done to be able to
+	// efficiently update the security rules concurrently, without having to
+	// block the request handlers for the time of the security rules update.
+	refCounter *atomic.Int32
+
+	// RWMutex protecting the R/W accesses to the internal rules data (stored
+	// in the handle).
+	mutex sync.RWMutex
+
+	// rulesetInfo holds information about rules initialization
+	rulesetInfo RulesetInfo
+}
+
+// NewHandle takes rules and the obfuscator config and returns a new WAF to work with. The order of action is the following:
+// - Open the WAF library libddwaf.so using dlopen
+// - Encode the rule free form object as a ddwaf_object to send them to the WAF
+// - Create a ddwaf_config object and fill the values
+// - Run ddwaf_init to get a handle from the waf
+// - Check for errors and streamline the ddwaf_ruleset_info returned
+func NewHandle(rules any, keyObfuscatorRegex string, valueObfuscatorRegex string) (*Handle, error) {
+
+	if err := InitWaf(); err != nil {
```

Comment:
Can we make `InitWaf` private?

Reply:
Yes

Reply:
Closing this for now as I have new ideas around Health vs InitWaf basically coming in my pending review.